### PR TITLE
Add sign in with publicKey

### PR DESCRIPTION
**Pull request by friedger**
This PR
- adds field `publicKey` to UserInputType
- adds `redirect` to sign in requests if the user input comes with a `publicKey` value. The `redirect` value is the encrypted redirect link.

This is a proof of concept towards 
https://github.com/opencollective/opencollective/issues/1749

Corresponding UI looks like this:
https://github.com/opencollective/opencollective-frontend/pull/1833

Probably it is not necessary to include blockstack.js as only function `encryptContent` is needed. 

After sign in the user name contains the blockstack id of the user:
Screenshot from 2019-05-27 11-07-28

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -27,6 +27,7 @@
     "babel-plugin-add-module-exports": "1.0.2",
     "babel-plugin-lodash": "3.3.4",
     "bcrypt": "3.0.6",
+    "blockstack": "^19.1.0",
```
Comment (znarf) [inline package.json]: You said integrating this `blockstack` library might not be mandatory. What would be the alternative? If it's simple, then let's do that.

File: server/controllers/users.js [hunk 2]
```diff
@@ -34,19 +59,25 @@ export const exists = async (req, res) => {
...
-      return emailLib.send('user.new.token', u.email, { loginLink }, { sendEvenIfNotProduction: true });
+      emailLib.send('user.new.token', u.email, { loginLink }, { sendEvenIfNotProduction: true });
+      if (user.publicKey) {
+        loginLink = blockstackLib.encryptLink(user.publicKey, loginLink);
+      }
     })
     .then(() => {
       const response = { success: true };
       // For e2e testing, we enable testuser+(admin|member)@opencollective.com to automatically receive the login link
-      if (process.env.NODE_ENV !== 'production' && user.email.match(/.*test.*@opencollective.com$/)) {
```
Comment (znarf) [inline server/controllers/users.js]: If there is no e2e tests, then let's not do that.

File: server/controllers/users.js [hunk 3]
```diff
@@ -34,19 +59,25 @@ export const exists = async (req, res) => {
...
+      if (
+        (process.env.NODE_ENV !== 'production' && user.email.match(/.*test.*@opencollective.com$/)) ||
+        user.publicKey
```
Comment (znarf) [inline server/controllers/users.js]: If there is no e2e tests, then let's not modify that.

File: server/controllers/users.js [hunk 4]
```diff
@@ -34,19 +59,25 @@ export const exists = async (req, res) => {
 export const signin = (req, res, next) => {
```
Comment (znarf) [inline server/controllers/users.js]: At this stage, can you have a specific `signin` endpoint for blockstack? And use a dedicated `exists` method?

To minimize footprint in `server/controllers/users.js`, supporting functions may be loaded from `lib/blockstack.js`

File: server/graphql/v1/mutations.js [hunk 5]
```diff
@@ -204,12 +206,25 @@ const mutations = {
           throw new Error('User already exists for given email');
         }
 
+        if (
```
Comment (znarf) [inline server/graphql/v1/mutations.js]: I don't think it's useful to do any modification in these GraphQL queries. The `signin` endpoint should be enough.

Comment (friedger) [reply server/graphql/v1/mutations.js]: This is for creating a new profile with blockstack.

Comment (znarf) [reply server/graphql/v1/mutations.js]: Ok, sorry, I was a bit wrong. `createUser` was indeed used for signups.

But, please handle that in a separate endpoint (ie: blockstackSignin / blockstackSignup), or a separated GraphQL mutation (ie: createUserWithBlockstack).

File: package.json (hunk 1 above)
Comment (friedger) [reply package.json]: I had a look but it is a bit more complicated.

File: server/graphql/v1/mutations.js (hunk 5 above)
Comment (friedger) [reply server/graphql/v1/mutations.js]: I have re-added it and moved everything into separate methods.

**Comment by friedger** [comment]
I have added the updateUserPublicKey mutation to better handle existing users

**Comment by friedger** [comment]
Can I do more to get this one merged?

**Comment by znarf** [comment]
@friedger nothing on the API side, this is more on the frontend side that we had open questions. We need to have another look and test again all the different cases.

**Comment by znarf** [comment]
This was pretty good work and I'm sorry we let this PR go stale.

Unfortunately, after discussing it with the team last week, there was no consensus to move it forward and I'll have to close it.

This could be revisited once supporting multiple identity providers become a priority.

**Pull request closed without merging**